cpp/src/kmeans/common.cuh

```diff
@@ -43,23 +45,55 @@
 
 #include <cuml/cluster/kmeans.hpp>
 
+#include <fstream>
 namespace ML {
 
 //@todo: Use GLOG once https://github.com/rapidsai/cuml/issues/100 is addressed.
-#define LOG(verbose, fmt, ...)                      \
-  do {                                              \
-    if (verbose) {                                  \
-      std::string msg;                              \
-      char verboseMsg[2048];                        \
-      std::sprintf(verboseMsg, fmt, ##__VA_ARGS__); \
-      msg += verboseMsg;                            \
-      std::cerr << msg;                             \
-    }                                               \
+#define LOG(handle, verbose, fmt, ...)                                   \
+  do {                                                                   \
+    bool verbose_ = verbose;                                             \
+    if (handle.commsInitialized()) {                                     \
+      const MLCommon::cumlCommunicator &comm = handle.getCommunicator(); \
+      const int my_rank = comm.getRank();                                \
+      verbose_ = verbose && (my_rank == 0);                              \
+    }                                                                    \
+    if (verbose_) {                                                      \
+      std::string msg;                                                   \
+      char verboseMsg[2048];                                             \
+      std::sprintf(verboseMsg, fmt, ##__VA_ARGS__);                      \
+      msg += verboseMsg;                                                 \
+      std::cerr << msg;                                                  \
+    }                                                                    \
   } while (0)
 
 namespace kmeans {
 namespace detail {
 
+template <typename LabelT, typename DataT>
+struct FusedL2NNReduceOp {
+  LabelT offset;
+
+  FusedL2NNReduceOp(LabelT _offset) : offset(_offset){};
+
+  typedef typename cub::KeyValuePair<LabelT, DataT> KVP;
+  DI void operator()(KVP *out, const KVP &other) {
+    if (other.value < out->value) {
+      out->key = offset + other.key;
+      out->value = other.value;
+    }
+  }
+
+  DI void init(KVP *out, DataT maxVal) {}
```

Comment:
@akkamesh I think the better approach would be to implement this function as expected, but have FusedL2NN prim expose an option to not initialize the output buffer. Let me provide you with such an interface, after which you can update this accordingly.

Reply:
Done